OpenRA.Mods.Common/Traits/ProductionStarport.cs

```diff
@@ -0,0 +1,144 @@
+	{
+		RallyPoint rp;
+		BulkProductionQueue queue;
+		Actor transport;
+
+		public ProductionStarport(ActorInitializer init, ProductionStarportInfo info)
+			: base(init, info) { }
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+
+			rp = self.TraitOrDefault<RallyPoint>();
+		}
+
+		public bool DeliverOrder(Actor producer, List<ActorInfo> orderedActors, string productionType)
+		{
+			if (IsTraitDisabled || IsTraitPaused)
+				return false;
+			var info = (ProductionStarportInfo)Info;
+			var owner = producer.Owner;
+			var map = owner.World.Map;
+			var waitTickbeforeSpawn = 15;
+			var aircraftInfo = producer.World.Map.Rules.Actors[info.ActorType].TraitInfo<AircraftInfo>();
+			queue = owner.World.ActorsWithTrait<BulkProductionQueue>().First(x => x.Actor.Owner == owner
+				&& x.Trait.GetActorsReadyForDelivery().Equals(orderedActors)).Trait;
+			CPos startPos;
+			WAngle spawnFacing;
+
+			startPos = producer.World.Map.ChooseClosestEdgeCell(producer.Location);
+			spawnFacing = producer.World.Map.FacingBetween(startPos, producer.Location, WAngle.Zero);
+
+			var exit = producer.Info.TraitInfos<ExitInfo>().First();
+
+			foreach (var tower in producer.TraitsImplementing<INotifyDelivery>())
+				tower.IncomingDelivery(producer);
+
+			owner.World.AddFrameEndTask(w =>
+			{
+				if (!producer.IsInWorld || producer.IsDead)
+				{
+					// Try to find another producer
+					var anotherProducer = queue.MostLikelyProducer();
+					if (anotherProducer.Trait != null && anotherProducer.Trait is ProductionStarport anotherStarport)
+						anotherStarport.DeliverOrder(anotherProducer.Actor, orderedActors, productionType);
+					else
+						queue.DeliverFinished();
+					return;
+				}
+
+				// aircrafts are delivered by themselfs
+				var exitCell = producer.Location + exit.ExitCell;
+				var destinations = rp != null && rp.Path.Count > 0 ? rp.Path : new List<CPos> { exitCell };
+
+				foreach (var orderedAircraft in orderedActors.Where(actor => actor.HasTraitInfo<AircraftInfo>()))
+				{
+					var altitude = orderedAircraft.TraitInfo<AircraftInfo>().CruiseAltitude;
+					var aircraft = w.CreateActor(orderedAircraft.Name, new TypeDictionary
+					{
+						new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WDist.Zero, WDist.Zero, altitude)),
+						new OwnerInit(owner),
+						new FacingInit(spawnFacing)
+					});
+					var move = aircraft.TraitOrDefault<IMove>();
+					if (move != null)
+					{
+						aircraft.QueueActivity(new Wait(waitTickbeforeSpawn));
+						waitTickbeforeSpawn += 15;
+
+						// first move must be to the Producer location
+						aircraft.QueueActivity(move.MoveTo(exitCell, 2, evaluateNearestMovableCell: true));
+						foreach (var cell in destinations)
+						{
+							aircraft.QueueActivity(move.MoveTo(cell, 2, evaluateNearestMovableCell: true));
+						}
+					}
+				}
+
+				orderedActors.RemoveAll(actor => actor.HasTraitInfo<AircraftInfo>());
+				waitTickbeforeSpawn = 0;
```

Comment:
```suggestion
				if (orderedActors.Count == 0)
				{
					queue.DeliverFinished();
					return;
				}
```

after ordered actors are removed we should check for wether any actors remain. If there's 0 then we should just exit

Reply:
Also remove waitTickbeforeSpawn = 0 as it is a variable on the stack. It will be removed as soon as the function goes out of scope so no need to reset it

you can also move its declaration into world end task so it's handled better